server/odc-service/src/main/java/com/oceanbase/odc/service/permission/database/DatabasePermissionService.java

```diff
@@ -42,32 +78,165 @@
+                permissionEntity.setType(PermissionType.PUBLIC_RESOURCE);
+                permissionEntity.setCreatorId(creatorId);
+                permissionEntity.setOrganizationId(organizationId);
+                permissionEntity.setBuiltIn(false);
+                permissionEntity.setExpireTime(expireTime);
+                permissionEntity.setAuthorizationType(AuthorizationType.USER_AUTHORIZATION);
+                permissionEntities.add(permissionEntity);
+            }
+        }
+        List<PermissionEntity> saved = permissionRepository.batchCreate(permissionEntities);
+        List<UserPermissionEntity> userPermissionEntities = new ArrayList<>();
+        for (PermissionEntity permissionEntity : saved) {
+            UserPermissionEntity userPermissionEntity = new UserPermissionEntity();
+            userPermissionEntity.setUserId(req.getUserId());
+            userPermissionEntity.setPermissionId(permissionEntity.getId());
+            userPermissionEntity.setCreatorId(creatorId);
+            userPermissionEntity.setOrganizationId(organizationId);
+            userPermissionEntities.add(userPermissionEntity);
+        }
+        userPermissionRepository.batchCreate(userPermissionEntities);
+        return saved.stream().map(PermissionEntity::getId).map(UserDatabasePermission::from)
+                .collect(Collectors.toList());
     }
 
     @Transactional(rollbackFor = Exception.class)
     @PreAuthenticate(hasAnyResourceRole = {"OWNER", "DBA"}, resourceType = "ODC_PROJECT", indexOfIdParam = 0)
-    public UserDatabasePermission revoke(@NotNull Long projectId, @NotNull Long permissionId) {
-        throw new NotImplementedException();
+    public List<UserDatabasePermission> batchRevoke(@NotNull Long projectId, @NotEmpty List<Long> ids) {
+        List<UserDatabasePermissionEntity> entities =
+                userDatabasePermissionRepository.findByProjectIdAndIdIn(projectId, ids);
+        List<Long> permissionIds =
+                entities.stream().map(UserDatabasePermissionEntity::getId).collect(Collectors.toList());
+        permissionRepository.deleteByIds(permissionIds);
+        userPermissionRepository.deleteByPermissionIds(permissionIds);
+        return permissionIds.stream().map(UserDatabasePermission::from).collect(Collectors.toList());
     }
 
-    @Transactional(rollbackFor = Exception.class)
-    @PreAuthenticate(hasAnyResourceRole = {"OWNER", "DBA", "DEVELOPER", "SECURITY_ADMINISTRATOR", "PARTICIPANT"},
-            resourceType = "ODC_PROJECT", indexOfIdParam = 0)
-    public UserDatabasePermission relinquish(@NotNull Long projectId, @NotNull Long permissionId) {
-        throw new NotImplementedException();
+    private Specification<UserDatabasePermissionEntity> getSpecificationByStatus(PermissionStatus status,
+            Date date) {
+        switch (status) {
+            case EXPIRED:
+                return UserDatabasePermissionSpec.expireTimeBefore(date);
+            case EXPIRING:
```

Comment:
There is overlap between the definitions of not-expired and expiring

Reply:
Yes, I refine it by change this enums to `ExpirationStatusFilter` for better readability.